File: windows/CodexTokenBar.Windows/RateLimitModels.cs
using System.Globalization;
using System.Text.Json;

namespace CodexTokenBar.Windows;

internal sealed record UsageWindow(double UsedPercent, int? WindowDurationMinutes, DateTimeOffset? ResetsAt)
{
    public double RemainingPercent => Math.Clamp(100 - UsedPercent, 0, 100);
}

internal sealed record UsageBucket(
    string Id,
    string? Name,
    UsageWindow? Primary,
    UsageWindow? Secondary,
    string? PlanType,
    bool? HasCredits,
    bool? UnlimitedCredits,
    string? CreditBalance,
    string? RateLimitReachedType)
{
    public UsageWindow? HeadlineWindow => Primary ?? Secondary;
    public string DisplayName => string.IsNullOrWhiteSpace(Name) ? Id : Name;
}

internal sealed record RateLimitResetCredit(
    string Id,
    string? Status,
    DateTimeOffset? ExpiresAt,
    string? Title);

internal sealed record UsageSnapshot(
    IReadOnlyList<UsageBucket> Buckets,
    int? AvailableResetCredits,
    IReadOnlyList<RateLimitResetCredit> ResetCredits,
    DateTimeOffset UpdatedAt)
{
    public UsageBucket? HeadlineBucket =>
        Buckets.FirstOrDefault(bucket => bucket.Id == "codex" && bucket.HeadlineWindow is not null)
        ?? Buckets.Where(bucket => bucket.HeadlineWindow is not null)
            .OrderBy(bucket => bucket.HeadlineWindow!.RemainingPercent)
            .FirstOrDefault();
}

internal static class RateLimitParser
{
    public static UsageSnapshot Parse(JsonElement root)
    {
        if (root.TryGetProperty("error", out var error) && error.ValueKind is not JsonValueKind.Null)
            throw new InvalidDataException(ServerError(error));

        var result = TryObject(root, "result") ?? root;
        var buckets = new List<UsageBucket>();
        var byId = TryObject(result, "rateLimitsByLimitId", "rate_limits_by_limit_id");

        if (byId is { } collection)
        {
            foreach (var property in collection.EnumerateObject())
            {
                if (property.Value.ValueKind != JsonValueKind.Object) continue;
                var bucket = ParseBucket(property.Value, property.Name);
                if (bucket is not null) buckets.Add(bucket);
            }
        }

        if (buckets.Count == 0)
        {
            var rawBucket = TryObject(result, "rateLimits", "rate_limits");
            if (rawBucket is { } value && ParseBucket(value, "codex") is { } bucket)
                buckets.Add(bucket);
        }

        if (buckets.Count == 0)
            throw new InvalidDataException("Codex 没有返回可显示的额度。");

        buckets.Sort((left, right) =>
        {
            if (left.Id == "codex") return right.Id == "codex" ? 0 : -1;
            if (right.Id == "codex") return 1;
            return StringComparer.CurrentCultureIgnoreCase.Compare(left.DisplayName, right.DisplayName);
        });

        int? availableCredits = null;
        var resetCreditDetails = new List<RateLimitResetCredit>();
        if (TryObject(result, "rateLimitResetCredits", "rate_limit_reset_credits") is { } credits)
        {
            availableCredits = Integer(credits, "availableCount", "available_count") is { } count
                ? Math.Max(0, count)
                : null;
            if (credits.TryGetProperty("credits", out var rawCredits) && rawCredits.ValueKind == JsonValueKind.Array)
            {
                foreach (var rawCredit in rawCredits.EnumerateArray())
                {
                    if (rawCredit.ValueKind != JsonValueKind.Object || Text(rawCredit, "id") is not { } id) continue;
                    DateTimeOffset? expiresAt = null;
                    if (Number(rawCredit, "expiresAt", "expires_at") is { } expiry)
                    {
                        try { expiresAt = DateTimeOffset.FromUnixTimeSeconds((long)expiry); }
                        catch (ArgumentOutOfRangeException) { }
                    }
                    resetCreditDetails.Add(new RateLimitResetCredit(
                        id,
                        Text(rawCredit, "status"),
                        expiresAt,
                        Text(rawCredit, "title")));
                }
            }
        }

        resetCreditDetails.Sort((left, right) => Nullable.Compare(left.ExpiresAt, right.ExpiresAt));
        return new UsageSnapshot(buckets, availableCredits, resetCreditDetails, DateTimeOffset.Now);
    }

    private static UsageBucket? ParseBucket(JsonElement raw, string fallbackId)
    {
        var primary = ParseWindow(TryObject(raw, "primary"));
        var secondary = ParseWindow(TryObject(raw, "secondary"));
        if (primary is null && secondary is null) return null;

        var credits = TryObject(raw, "credits");
        return new UsageBucket(
            Text(raw, "limitId", "limit_id") ?? fallbackId,
            Text(raw, "limitName", "limit_name"),
            primary,
            secondary,
            Text(raw, "planType", "plan_type"),
            credits is { } creditObject ? Boolean(creditObject, "hasCredits", "has_credits") : null,
            credits is { } unlimitedObject ? Boolean(unlimitedObject, "unlimited") : null,
            credits is { } balanceObject ? Text(balanceObject, "balance") : null,
            Text(raw, "rateLimitReachedType", "rate_limit_reached_type"));
    }

    private static UsageWindow? ParseWindow(JsonElement? raw)
    {
        if (raw is not { } value || Number(value, "usedPercent", "used_percent") is not { } used)
            return null;

        DateTimeOffset? resetsAt = null;
        if (Number(value, "resetsAt", "resets_at") is { } timestamp)
        {
            try { resetsAt = DateTimeOffset.FromUnixTimeSeconds((long)timestamp); }
            catch (ArgumentOutOfRangeException) { }
        }

        return new UsageWindow(
            used,
            Integer(value, "windowDurationMins", "window_duration_mins", "window_minutes"),
            resetsAt);
    }

    private static JsonElement? TryObject(JsonElement value, params string[] names)
    {
        foreach (var name in names)
            if (value.ValueKind == JsonValueKind.Object &&
                value.TryGetProperty(name, out var child) && child.ValueKind == JsonValueKind.Object)
                return child;
        return null;
    }

    private static string? Text(JsonElement value, params string[] names)
    {
        foreach (var name in names)
        {
            if (!value.TryGetProperty(name, out var child)) continue;
            var result = child.ValueKind switch
            {
                JsonValueKind.String => child.GetString(),
                JsonValueKind.Number => child.GetRawText(),
                _ => null
            };
            if (!string.IsNullOrWhiteSpace(result)) return result.Trim();
        }
        return null;
    }

    private static double? Number(JsonElement value, params string[] names)
    {
        foreach (var name in names)
        {
            if (!value.TryGetProperty(name, out var child)) continue;
            if (child.ValueKind == JsonValueKind.Number && child.TryGetDouble(out var number)) return number;
            if (child.ValueKind == JsonValueKind.String &&
                double.TryParse(child.GetString(), NumberStyles.Float, CultureInfo.InvariantCulture, out number))
                return number;
        }
        return null;
    }

    private static int? Integer(JsonElement value, params string[] names) =>
        Number(value, names) is { } number ? (int)number : null;

    private static bool? Boolean(JsonElement value, params string[] names)
    {
        foreach (var name in names)
        {
            if (!value.TryGetProperty(name, out var child)) continue;
            if (child.ValueKind == JsonValueKind.True) return true;
            if (child.ValueKind == JsonValueKind.False) return false;
            if (child.ValueKind == JsonValueKind.String && bool.TryParse(child.GetString(), out var parsed))
                return parsed;
        }
        return null;
    }

    private static string ServerError(JsonElement error) =>
        Text(error, "message") ?? "Codex 返回了未知错误。";
}

internal sealed record DailyTokenUsage(string StartDate, long Tokens);

internal sealed record TokenUsageSnapshot(IReadOnlyList<DailyTokenUsage> DailyUsage, long? LifetimeTokens)
{
    public string? LatestStartDate => DailyUsage.LastOrDefault()?.StartDate;
    public long? TokensOn(string startDate) => DailyUsage.FirstOrDefault(item => item.StartDate == startDate)?.Tokens;
}

internal static class TokenUsageParser
{
    public static TokenUsageSnapshot Parse(JsonElement root)
    {
        if (root.TryGetProperty("error", out var error) && error.ValueKind != JsonValueKind.Null)
            throw new InvalidDataException("Codex 返回了无法识别的 token 用量数据。");

        var result = root.TryGetProperty("result", out var resultElement) && resultElement.ValueKind == JsonValueKind.Object
            ? resultElement
            : root;
        long? lifetime = null;
        if (result.TryGetProperty("summary", out var summary) && summary.ValueKind == JsonValueKind.Object)
            lifetime = NonnegativeInteger(summary, "lifetimeTokens", "lifetime_tokens");

        var byDate = new Dictionary<string, DailyTokenUsage>();
        JsonElement rawBuckets = default;
        var hasBuckets = result.TryGetProperty("dailyUsageBuckets", out rawBuckets) ||
                         result.TryGetProperty("daily_usage_buckets", out rawBuckets);
        if (hasBuckets && rawBuckets.ValueKind == JsonValueKind.Array)
        {
            foreach (var raw in rawBuckets.EnumerateArray())
            {
                if (raw.ValueKind != JsonValueKind.Object) continue;
                var startDate = String(raw, "startDate", "start_date");
                var tokens = NonnegativeInteger(raw, "tokens");
                if (startDate is not null && tokens is not null)
                    byDate[startDate] = new DailyTokenUsage(startDate, tokens.Value);
            }
        }
        return new TokenUsageSnapshot(byDate.Values.OrderBy(item => item.StartDate).ToArray(), lifetime);
    }

    private static string? String(JsonElement value, params string[] names)
    {
        foreach (var name in names)
            if (value.TryGetProperty(name, out var child) && child.ValueKind == JsonValueKind.String &&
                !string.IsNullOrWhiteSpace(child.GetString()))
                return child.GetString()!.Trim();
        return null;
    }

    private static long? NonnegativeInteger(JsonElement value, params string[] names)
    {
        foreach (var name in names)
        {
            if (!value.TryGetProperty(name, out var child)) continue;
            long parsed;
            if (child.ValueKind == JsonValueKind.Number && child.TryGetInt64(out parsed) && parsed >= 0) return parsed;
            if (child.ValueKind == JsonValueKind.String && long.TryParse(child.GetString(), out parsed) && parsed >= 0) return parsed;
        }
        return null;
    }
}

File: Sources/CodexTokenCore/RateLimitModels.swift
import Foundation

public struct UsageWindow: Equatable, Sendable {
    public let usedPercent: Double
    public let windowDurationMinutes: Int?
    public let resetsAt: Date?

    public init(usedPercent: Double, windowDurationMinutes: Int?, resetsAt: Date?) {
        self.usedPercent = usedPercent
        self.windowDurationMinutes = windowDurationMinutes
        self.resetsAt = resetsAt
    }

    public var remainingPercent: Double {
        min(100, max(0, 100 - usedPercent))
    }
}

public struct UsageBucket: Equatable, Sendable {
    public let id: String
    public let name: String?
    public let primary: UsageWindow?
    public let secondary: UsageWindow?
    public let planType: String?
    public let hasCredits: Bool?
    public let unlimitedCredits: Bool?
    public let creditBalance: String?
    public let rateLimitReachedType: String?

    public init(
        id: String,
        name: String?,
        primary: UsageWindow?,
        secondary: UsageWindow?,
        planType: String?,
        hasCredits: Bool?,
        unlimitedCredits: Bool?,
        creditBalance: String?,
        rateLimitReachedType: String?
    ) {
        self.id = id
        self.name = name
        self.primary = primary
        self.secondary = secondary
        self.planType = planType
        self.hasCredits = hasCredits
        self.unlimitedCredits = unlimitedCredits
        self.creditBalance = creditBalance
        self.rateLimitReachedType = rateLimitReachedType
    }

    public var headlineWindow: UsageWindow? {
        primary ?? secondary
    }
}

public struct RateLimitResetCredit: Equatable, Sendable {
    public let id: String
    public let resetType: String?
    public let status: String?
    public let grantedAt: Date?
    public let expiresAt: Date?
    public let title: String?
    public let description: String?

    public init(
        id: String,
        resetType: String?,
        status: String?,
        grantedAt: Date?,
        expiresAt: Date?,
        title: String?,
        description: String?
    ) {
        self.id = id
        self.resetType = resetType
        self.status = status
        self.grantedAt = grantedAt
        self.expiresAt = expiresAt
        self.title = title
        self.description = description
    }
}

public struct UsageSnapshot: Equatable, Sendable {
    public let buckets: [UsageBucket]
    public let availableResetCredits: Int?
    public let resetCredits: [RateLimitResetCredit]
    public let updatedAt: Date

    public init(
        buckets: [UsageBucket],
        availableResetCredits: Int?,
        resetCredits: [RateLimitResetCredit] = [],
        updatedAt: Date
    ) {
        self.buckets = buckets
        self.availableResetCredits = availableResetCredits
        self.resetCredits = resetCredits
        self.updatedAt = updatedAt
    }

    public var headlineBucket: UsageBucket? {
        if let codex = buckets.first(where: { $0.id == "codex" && $0.headlineWindow != nil }) {
            return codex
        }

        return buckets
            .filter { $0.headlineWindow != nil }
            .min {
                ($0.headlineWindow?.remainingPercent ?? 100) <
                    ($1.headlineWindow?.remainingPercent ?? 100)
            }
    }
}

public enum ResetCountdownFormatter {
    public static func string(until resetDate: Date, now: Date = Date()) -> String {
        let remainingSeconds = resetDate.timeIntervalSince(now)

        guard remainingSeconds > 0 else {
            return "正在重置"
        }

        let totalHours = max(1, Int(ceil(remainingSeconds / 3_600)))
        let days = totalHours / 24
        let hours = totalHours % 24

        if days > 0, hours > 0 {
            return "还有 \(days) 天 \(hours) 小时"
        }
        if days > 0 {
            return "还有 \(days) 天"
        }
        return "还有 \(hours) 小时"
    }
}

public enum DailyAllowanceCalculator {
    public static func percentPerDay(
        remainingPercent: Double,
        until resetDate: Date,
        now: Date = Date()
    ) -> Double? {
        let remainingSeconds = resetDate.timeIntervalSince(now)
        guard remainingSeconds > 0, remainingPercent > 0 else {
            return nil
        }

        let remainingDays = remainingSeconds / 86_400
        return remainingPercent / remainingDays
    }
}

public enum RateLimitParseError: LocalizedError {
    case invalidEnvelope
    case noRateLimits

    public var errorDescription: String? {
        switch self {
        case .invalidEnvelope:
            return "Codex 返回了无法识别的数据。"
        case .noRateLimits:
            return "Codex 没有返回可显示的额度。"
        }
    }
}

public enum RateLimitParser {
    public static func parseResponse(data: Data, now: Date = Date()) throws -> UsageSnapshot {
        guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw RateLimitParseError.invalidEnvelope
        }
        guard root["error"] == nil else { throw RateLimitParseError.invalidEnvelope }

        let result = (root["result"] as? [String: Any]) ?? root
        return try parseResult(result, now: now)
    }

    public static func parseResult(_ result: [String: Any], now: Date = Date()) throws -> UsageSnapshot {
        var buckets: [UsageBucket] = []

        if let byID = dictionary(result["rateLimitsByLimitId"] ?? result["rate_limits_by_limit_id"]) {
            for (fallbackID, value) in byID {
                guard let rawBucket = dictionary(value) else { continue }
                if let bucket = parseBucket(rawBucket, fallbackID: fallbackID) {
                    buckets.append(bucket)
                }
            }
        }

        if buckets.isEmpty,
           let rawBucket = dictionary(result["rateLimits"] ?? result["rate_limits"]),
           let bucket = parseBucket(rawBucket, fallbackID: "codex") {
            buckets.append(bucket)
        }

        guard !buckets.isEmpty else {
            throw RateLimitParseError.noRateLimits
        }

        buckets.sort { lhs, rhs in
            if lhs.id == "codex" { return rhs.id != "codex" }
            if rhs.id == "codex" { return false }
            return displayName(for: lhs).localizedCaseInsensitiveCompare(displayName(for: rhs)) == .orderedAscending
        }

        let resetCredits = dictionary(result["rateLimitResetCredits"] ?? result["rate_limit_reset_credits"])
        let availableCount = integer(
            resetCredits?["availableCount"] ?? resetCredits?["available_count"]
        ).map { max(0, $0) }
        var resetCreditDetails = array(resetCredits?["credits"])
            .compactMap(dictionary)
            .compactMap(parseResetCredit)
        resetCreditDetails.sort(by: resetCreditSort)

        return UsageSnapshot(
            buckets: buckets,
            availableResetCredits: availableCount,
            resetCredits: resetCreditDetails,
            updatedAt: now
        )
    }

    private static func parseResetCredit(_ raw: [String: Any]) -> RateLimitResetCredit? {
        guard let id = nonEmptyString(raw["id"]) else { return nil }

        let grantedTimestamp = number(raw["grantedAt"] ?? raw["granted_at"])
        let expiryTimestamp = number(raw["expiresAt"] ?? raw["expires_at"])

        return RateLimitResetCredit(
            id: id,
            resetType: nonEmptyString(raw["resetType"] ?? raw["reset_type"]),
            status: nonEmptyString(raw["status"]),
            grantedAt: grantedTimestamp.map(Date.init(timeIntervalSince1970:)),
            expiresAt: expiryTimestamp.map(Date.init(timeIntervalSince1970:)),
            title: nonEmptyString(raw["title"]),
            description: nonEmptyString(raw["description"])
        )
    }

    private static func resetCreditSort(
        _ lhs: RateLimitResetCredit,
        _ rhs: RateLimitResetCredit
    ) -> Bool {
        switch (lhs.expiresAt, rhs.expiresAt) {
        case let (lhsDate?, rhsDate?):
            if lhsDate != rhsDate { return lhsDate < rhsDate }
        case (_?, nil):
            return true
        case (nil, _?):
            return false
        case (nil, nil):
            break
        }
        return lhs.id < rhs.id
    }

    private static func parseBucket(_ raw: [String: Any], fallbackID: String) -> UsageBucket? {
        let id = string(raw["limitId"] ?? raw["limit_id"]) ?? fallbackID
        let primary = parseWindow(dictionary(raw["primary"]))
        let secondary = parseWindow(dictionary(raw["secondary"]))

        guard primary != nil || secondary != nil else { return nil }

        let credits = dictionary(raw["credits"])
        return UsageBucket(
            id: id,
            name: nonEmptyString(raw["limitName"] ?? raw["limit_name"]),
            primary: primary,
            secondary: secondary,
            planType: nonEmptyString(raw["planType"] ?? raw["plan_type"]),
            hasCredits: boolean(credits?["hasCredits"] ?? credits?["has_credits"]),
            unlimitedCredits: boolean(credits?["unlimited"]),
            creditBalance: string(credits?["balance"]),
            rateLimitReachedType: nonEmptyString(raw["rateLimitReachedType"] ?? raw["rate_limit_reached_type"])
        )
    }

    private static func parseWindow(_ raw: [String: Any]?) -> UsageWindow? {
        guard let raw, let usedPercent = number(raw["usedPercent"] ?? raw["used_percent"]) else {
            return nil
        }

        let duration = integer(
            raw["windowDurationMins"] ?? raw["window_duration_mins"] ?? raw["window_minutes"]
        )
        let resetTimestamp = number(raw["resetsAt"] ?? raw["resets_at"])

        return UsageWindow(
            usedPercent: usedPercent,
            windowDurationMinutes: duration,
            resetsAt: resetTimestamp.map(Date.init(timeIntervalSince1970:))
        )
    }

    private static func displayName(for bucket: UsageBucket) -> String {
        bucket.name ?? bucket.id
    }

    private static func dictionary(_ value: Any?) -> [String: Any]? {
        value as? [String: Any]
    }

    private static func array(_ value: Any?) -> [Any] {
        value as? [Any] ?? []
    }

    private static func string(_ value: Any?) -> String? {
        if let value = value as? String { return value }
        if let value = value as? NSNumber { return value.stringValue }
        return nil
    }

    private static func nonEmptyString(_ value: Any?) -> String? {
        guard let value = string(value)?.trimmingCharacters(in: .whitespacesAndNewlines), !value.isEmpty else {
            return nil
        }
        return value
    }

    private static func number(_ value: Any?) -> Double? {
        if let value = value as? NSNumber { return value.doubleValue }
        if let value = value as? String { return Double(value) }
        return nil
    }

    private static func integer(_ value: Any?) -> Int? {
        number(value).map(Int.init)
    }

    private static func boolean(_ value: Any?) -> Bool? {
        if let value = value as? Bool { return value }
        if let value = value as? NSNumber { return value.boolValue }
        if let value = value as? String {
            switch value.lowercased() {
            case "true", "1", "yes": return true
            case "false", "0", "no": return false
            default: return nil
            }
        }
        return nil
    }
}
